src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php

```diff
@@ -15,13 +15,16 @@
 use Illuminate\Support\Str;
 use InvalidArgumentException;
 
-/** @mixin \Illuminate\Database\Eloquent\Builder */
+/**
+ * @template TModel of \Illuminate\Database\Eloquent\Model
+ * @mixin \Illuminate\Database\Eloquent\Builder<TModel>
+ */
 trait QueriesRelationships
 {
     /**
      * Add a relationship count / exists condition to the query.
      *
-     * @param  \Illuminate\Database\Eloquent\Relations\Relation<*, *, *>|string  $relation
+     * @param  \Illuminate\Database\Eloquent\Relations\Relation<*, *, *>|\Closure(TModel): \Illuminate\Database\Eloquent\Relations|string  $relation
```

Comment:
You need to specify the base Relation class, and it's more readable for the closure to be surrounded by parenthesis as it's in the middle of a union:

Here's how I would write this:

```suggestion
     * @param  (\Closure(TModel): \Illuminate\Database\Eloquent\Relations\Relation<*, *, *>)|\Illuminate\Database\Eloquent\Relations\Relation<*, *, *>|string  $relation
```


Also, please add type tests for each of these methods to `types/Database/Eloquent/Builder.php`, for example:

```php
    assertType('Illuminate\Database\Eloquent\Builder<User>', $query->has(function ($user) {
        assertType('User', $user);
    });
```